src/test/java/com/microsoft/sqlserver/jdbc/requestboundary/RequestBoundaryMethodsTest.java

```diff
@@ -0,0 +1,421 @@
+                con.endRequest();
+
+                assertTrue(stmt.isClosed());
+                assertTrue(ps.isClosed());
+                assertTrue(stmt1.isClosed());
+                assertTrue(cs.isClosed());
+                assertTrue(rs1.isClosed());
+            }
+        }
+        finally {
+            if (null != stmt) {
+                stmt.close();
+            }
+            if (null != stmt1) {
+                stmt1.close();
+            }
+            if (null != ps) {
+                ps.close();
+            }
+            if (null != cs) {
+                cs.close();
+            }
+            if (null != con) {
+                con.close();
+            }
+        }
+    }
+
+    /**
+     * Tests Request Boundary methods in a multi-threaded environment.
+     * 
+     * @throws SQLException
+     */
+    @Test
+    public void testThreads() throws SQLException {
+        try {
+            con = connect();
+            if (Utils.isJDBC43AndGreater(con)) {
+                Thread thread1 = new Thread() {
+                    public void run() {
+                        try {
+                            con.setNetworkTimeout(null, 100);
+                            con.setHoldability(ResultSet.CLOSE_CURSORS_AT_COMMIT);
+                        }
+                        catch (SQLException e) {
+                            e.printStackTrace();
+                        }
+                    }
+                };
+
+                Thread thread2 = new Thread() {
+                    public void run() {
+                        try {
+                            stmt = con.createStatement();
+                            ResultSet rs = stmt.executeQuery("SELECT 1");
+                            rs.next();
+                            assertEquals(1, rs.getInt(1));
+                        }
+                        catch (SQLException e) {
+                            e.printStackTrace();
+                        }
+                    }
+                };
+
+                Thread thread3 = new Thread() {
+                    public void run() {
+                        try {
+                            pstmt = con.prepareStatement("SELECT 1");
+                            ResultSet rs = pstmt.executeQuery();
+                            rs.next();
+                            assertEquals(1, rs.getInt(1));
+                        }
+                        catch (SQLException e) {
+                            e.printStackTrace();
+                        }
+
+                    }
+                };
+
+                int originalNetworkTimeout = con.getNetworkTimeout();
+                int originalHoldability = con.getHoldability();
+                con.beginRequest();
+                thread1.start();
+                thread2.start();
+                thread3.start();
+                try {
+                    // Wait for threads to complete
+                    Thread.sleep(3000);

```

Comment:
Again, Thread.sleep is not required. Java Concurrency has `CountDownLatch.await()` or `ExecutorService.awaitTermination()` APIs that would wait for all threads to finish and then proceed further.

Reply:
Agreed.